File: basics/articulation_points.cpp
#include <bits/stdc++.h>
using namespace std;
const int inf = 0x3f3f3f3f;
int N, M, a, b, parent[200000], low[200000], disc[200000], timee = 0;
bool vis[200000], ap[200000];
vector<int> adj[200000];


void dfs(int curr){
  vis[curr] = true;
  timee++;
  disc[curr] = timee;
  low[curr] = timee;
  int children = 0;
  for (auto node: adj[curr]){
    if (!vis[node]){
      children++;
      parent[node] = curr;
      dfs(node);
      low[curr] = min(low[curr], low[node]);
      if (parent[curr] == -1 && children > 1){
        ap[curr] = true;
      } if (parent[curr] != -1 && low[node] >= disc[curr]){
        ap[curr] = true;
      }
    }else if (parent[curr] != node){
      low[curr] = min(low[curr], disc[node]);
    }
  }
}


int main(){
  memset(vis, false, sizeof(vis));
  memset(ap, false, sizeof(ap));
  memset(parent, -1, sizeof(parent));
  memset(low, inf, sizeof(low));
  memset(disc, 0, sizeof(disc));

  cin >> N >> M;
  for (int i = 0; i < M; i++){
    cin >> a >> b;
    adj[a].push_back(b);
    adj[b].push_back(a);
  }

  for (int i = 1; i <= N; i++){
    if (!vis[i]){
      dfs(i);
    }
  }

  for (int i = 1; i <= N; i++){
    cout << ap[i] << " ";
  } cout << "\n";

  return 0;
}

File: basics/articulation_points.py
import sys
N, M = [int(x) for x in sys.stdin.readline().split()]
adj = [[] for x in range(N+1)]
for i in range(M):
  a, b = [int(x) for x in sys.stdin.readline().split()]
  adj[a].append(b)
  adj[b].append(a)

time = 0
disc = [0 for x in range(N+1)]
parent = [-1 for x in range(N+1)]
low = [float("inf") for x in range(N+1)]
ap = [False for x in range(N+1)]
vis = [False for x in range(N+1)]


def dfs(curr):
  global time
  vis[curr] = True
  time += 1
  disc[curr] = time 
  low[curr] = time
  children = 0
  for node in adj[curr]:
    if not vis[node]:
      children += 1
      parent[node] = curr
      dfs(node)
      low[curr] = min(low[curr], low[node])
      if parent[curr] == -1 and children > 1:
        ap[curr] = True
      if parent[curr] != -1 and low[node] >= disc[curr]:
        ap[curr] = True

    elif parent[curr] != node:
      low[curr] = min(low[curr], disc[node])


for i in range(1, N+1):
  if not vis[i]:
    dfs(i)

print(ap)